[Pattern Draft] Internal Developer Platform (#720)
gruetter/innerSourceTest: patterns/1-initial/internal-developer-platform.md

- **Pluggable Architecture**: The platform must be extensible to adapt to changing organizational needs. It should support integration with various third-party services, tools, and APIs via a plugin architecture to allow teams to extend functionality as required​.

- **InnerSource Practices**: Encourage InnerSource practices within the IDP to enhance collaboration and knowledge sharing between teams. By facilitating code reuse and contribution across the organization, InnerSource reduces duplication and promotes a culture of transparency​.

- **Automation & Infrastructure as Code (IaC)**: Automate repetitive tasks such as environment provisioning, testing, and deployment. Integrating Infrastructure as Code (IaC) ensures that infrastructure is treated like code, versioned, and managed in a scalable and reliable manner​

## Resulting Context

Adoption of an Internal Developer Platform results in:

> - Stremline development workflow and improved developer efficiency through self-service workflows and standardized environments.

Reviewer: ```suggestion
- Streamline development workflow and improved developer efficiency through self-service workflows and standardized environments.
```

> - Stremline development workflow and improved developer efficiency through self-service workflows and standardized environments.

Reviewer 2: typo: Stremline --> Streamline

Reviewer: ```suggestion
- Streamline development workflow and improved developer efficiency through self-service workflows and standardized environments.
```

Author: Thanks for highlighting. Corrected with the last commit.

## Title

Internal Developer Platform

## Patlet

As InnerSource adoption increases throughout an organisation, it is not unusual that project teams start to face inefficiencies in scaling their efforts due to fragmented tooling, environments, and workflows. An Internal Developer Platform (IDP) provides a way to tackle this type of challenges through a centralized, self-service system that standardizes development environments and integrates tools to enhance consistency, collaboration, and developer productivity.

> There is a symbiotic relationship between the IDP and InnerSource as their relationship enhances the scalability and effectiveness of both practices, especially in large organizations.

Reviewer: While this seems a bit too long for a Patlet, I suggest to leave it as is for now. We can adapt this once we add more Known Instances and feedback from other orgs, before making this pattern available in our online book. (maturity level "Structured")

### InnerSource Practices Amplify the Adoption of IDP

**Channeling Contributions to Platform Development**: InnerSource creates an avenue for developers across the organization to contribute directly to the IDP. For example, they can suggest or implement features that address common pain points or enhance usability. This community-driven approach ensures that the IDP evolves in response to real-world challenges faced by its users, making it more relevant and effective.

**Encouraging Shared Ownership of the Platform**: Through InnerSource, developers treat the IDP as a shared product rather than something imposed by a centralized team. This increases adoption and engagement, as teams feel a sense of ownership and responsibility for its success.

**Improving Documentation and Knowledge Sharing**: InnerSource practices naturally encourage contributions to documentation. Teams using the IDP can provide feedback, improve documentation, or add usage examples, ensuring that knowledge is continuously refined and up-to-date.

> 

Reviewer: The Palet should stay pretty short (max 2 sentences describing problem and solution). Like the shortest possible summary of what this pattern is about.

Therefore we should move this extra text somewhere else in the pattern.

For the time being you could move this to the "Rationale" section, what do you think?

Without a standardized foundation, InnerSource efforts struggle to scale as inconsistencies across teams lead to inefficiencies, delays, and misaligned governance practices. Contributors face challenges in understanding and integrating with unfamiliar environments, while teams resist external contributions due to perceived risks or integration overhead. An Internal Developer Platform (IDP) mitigates these issues by providing a self-service, unified platform that standardizes tools, processes, and environments. This enables InnerSource contributors to focus on delivering value rather than overcoming technical incompatibilities, fostering a culture of collaboration and accelerating the adoption of InnerSource at scale.

> ## Story (optional)
> 
> (to be done)
> 

Reviewer: ```suggestion
```

## Context

> This pattern emerges in organizations with multiple InnerSource development teams working on different projects. As teams grow and evolve, discrepancies in the setup of development environments and access to resources become more prominent, hampering collaboration, introducing security risks, and reducing developer productivity. The need for uniformity without compromising flexibility drives the adoption of an IDP.

Reviewer: I broke these up into bullets, so that readers can use this like a checklist.

I did some rewording, however I tried to keep the intended meaning. Please check if it worked :)

```suggestion
- The organization has development teams working on different projects. These teams are already using InnerSource practices.
- As teams grow and evolve, discrepancies in the setup of development environments and access to resources become more prominent, hampering collaboration, introducing security risks, and reducing developer productivity.
- The teams (as well as the leadership of the organization) are looking to increase consistency in tools and processes without compromising on the flexibility to adapt to project-specific needs.
```

Question:

What do you mean by "access to resources"? An example could help here.

Reviewer: In your ISC community talk you mention that 99% of the work has to happen before setting up the IDP itself. i.e. self-service functionality for the developers has to be developed first, before it can be exposed through an IDP.

Therefore I think we should capture these pre-conditions here, in the form of additional elements of the Context section.

For example:

- the organization has a DevEx team (aka platform development team), tasked to improve developer productivity by reducing friction from suboptimal tools and processes
- at least some self-service style functionality has already been developed by the DevEx team

**Building actionable documentation**: An IDP integrates documentation directly into the development workflows via self-service portals. Teams can access API documentation, code examples, and templates from the same place they access tools and environments. This centralization ensures that documentation is kept up-to-date and easily accessible. Furthermore, by encouraging direct InnerSource contributions into the documentation itself (from users), teams can continuously improve documentation based on real-world usage​.

## Rationale

An IDP provides a solution to the competing forces of consistency, flexibility, and scalability. By centralizing the management of development environments and tooling while offering self-service capabilities, it balances the needs of both developers and platform teams. InnerSource practices further contribute by fostering a culture of collaboration and reuse across the organization​. It supports scalable and controlled growth within the internal software development community, facilitating the adoption of wider InnerSource practices.

## Known Instances (optional)

(known instances to be added)

## Status

> Initial Draft

Reviewer: ```suggestion
* Initial
```

## Author(s)

Vincent Caldeira

## Acknowledgments (optional)

Include those who assisted in helping with this pattern - both for attribution and for possible future follow up.
Though optional, most patterns should list who helped in their creation.

> 

Reviewer 3: It could be good add a `References` chapter to link this pattern with others that could be related, such as:

* [InnerSource Portal](https://patterns.innersourcecommons.org/p/innersource-portal)

Maybe others could be listed here too. Ideas?

Reviewer: Sure, why not. I think we already have some other patterns that have used a References section like this.

You can also include links to other patterns as part of the regular text. i.e. if there is an area in this pattern where we naturally talk about the relationship between IDP and InnerSource Portal, we could add the link there.

Reviewer: @rmarting I am trying to get this pattern merged, so that it can more easily be linked to. Therefore we might not add this link to the InnerSource Portal now.

However we can leave this comment thread open if you like, so that we might remember it when we improve the pattern in the future (e.g. when we add Known Instances etc).

## Alias (optional)

> (known aliases to be added)

Reviewer: ```suggestion
## Links

* [Promoting InnerSource Practices with an Internal Developer Platform](https://www.youtube.com/watch?v=mqeBbQcDjTE&ab_channel=InnerSourceCommons)
* [internaldeveloperplatform.org](https://internaldeveloperplatform.org/)
```

We can remove this section, as the pattern does not have aliases (alias = different name for the same pattern).

Instead I am suggesting to add a section that links to other material that helps to understand this pattern better, like the ISC community call related to this pattern etc.

**Building actionable documentation**: An IDP integrates documentation directly into the development workflows via self-service portals. Teams can access API documentation, code examples, and templates from the same place they access tools and environments. This centralization ensures that documentation is kept up-to-date and easily accessible. Furthermore, by encouraging direct InnerSource contributions into the documentation itself (from users), teams can continuously improve documentation based on real-world usage​.

## Rationale

An IDP provides a solution to the competing forces of consistency, flexibility, and scalability. By centralizing the management of development environments and tooling while offering self-service capabilities, it balances the needs of both developers and platform teams. InnerSource practices further contribute by fostering a culture of collaboration and reuse across the organization​. It supports scalable and controlled growth within the internal software development community, facilitating the adoption of wider InnerSource practices.

> ## Known Instances (optional)

Reviewer: ```suggestion
## Known Instances
```

(known instances to be added)

## Status

* Initial

## Author(s)

Vincent Caldeira

## Acknowledgments (optional)

> Include those who assisted in helping with this pattern - both for attribution and for possible future follow up.
> Though optional, most patterns should list who helped in their creation.

Reviewer: ```suggestion
- [Jose Roman Martin Gil](https://github.com/rmarting) for providing early feedback on this pattern and helping to look for organizations that have implemented this pattern
```
